### ISSUE TITLE
Problem reading files when looping in parallel

### ISSUE THREAD

**Issue by drnickisaac** [BiologicalRecordsCentre/wrappeR#19]
I've run into problem when reading files from the ObjectStore in parallel. The code below runs in about 5-6 minutes (on Datalabs). I've run it twice and both times 39 of the 205 species fail to load. When digging into the output, the error is
`"error in load(paste0(indata, species, \".rdata\")) : \n  empty (zero-byte) input file\n"`

In both cases the first species to fail is the third one. I haven't compared the locations of the other 38 fails, but it feels safe to assume that the same files are causing th problem each time.

NB running in serial does not have this problem (but fails elsewhere!).
NB this is a pretty rare event. I've successfully loaded 28 of Charlie's groups and only Hoverflies and LeafSeedBeetles experienced a problem.

```
taxa_to_include <- "Hoverflies"
(n_taxa <- length(taxa_to_include))

######################## Create roster
## set up a roster specifying what filters should be applied in call to applyFilters
rosterTEMP <- wrappeR::createRoster(index = n_taxa,
                                modPath = rep("/data-s3/occmods/", n_taxa), 
                                metaPath = rep("/data-s3/metadata/", n_taxa),
                                ver = rep("2017_Charlie", n_taxa),
                                indicator = rep("all", n_taxa),
                                region = rep("ENGLAND", n_taxa),
                                nSamps = rep(10, n_taxa),
                                minObs = rep(100, n_taxa),
                                write = rep(FALSE, n_taxa), 
                                outPath = rep(0, n_taxa),
                                clipBy = rep("species", n_taxa), ## here you specify "species" or "group"
                                group = taxa_to_include,
                                t0 = rep(1970, n_taxa),
                                tn = rep(2015, n_taxa))


################################### # READ in the models
system.time(
  sampDat_HOV_PAR <-  lapply(rosterTEMP, 
                     wrappeR::applySamp,
                     parallel=TRUE, sample=TRUE)
)
```

**Comment by AugustT** [comment]
This sounds like somethin Mark can explore with Ola and Ian. @robboyd , can you add Mark as a collaborator so I can assign to him?

**Comment by AugustT** [comment]
Can you expand on the serial problem?

**Comment by robboyd** [comment]
@AugustT  have invited Mark just need him to accept. @drnickisaac Rob had this problem too and we think there's a problem with a small number of the model output files on datalabs.

**Comment by drnickisaac** [comment]
@robboyd  I am skeptical that it's a problem with the files on ObjectStore. The file in the example above works fine when read in serial.

**Comment by drnickisaac** [comment]
@robboyd  Clarification: the error I'm reporting is not associated with files on Object Store. There may be other issues with the files on ObjectStore.
@AugustT the serial error popped the following message:
```
#Error in load(paste0(indata, species, ".rdata")) : 
#  empty (zero-byte) input file
```
The last species that loaded successfully was "Platycheirus peltatus".  The failure would be either the next species (alphabetically), or the next species with 50 records. If there is a problem with zero-byte files then it should be quite easy to a) catch the error and b) identify which species and update them with the uncorrupted file.
But none of this solves my problem when running in parallel.

**Comment by drnickisaac** [comment]
Update: some further observations:
1) Across 29 groups, there were errors in just two taxa: Hoverflies & LeafSeedBeetles. In both cases there were 39 species that failed, although the total number is not the same (205 hoverflies, 135 LeafSeedBeetles). 
2) the same number of species (39) in the same two groups returned errors when reading just the `a` parameters, not doing the sample. This implies it's some kind of memory issue, rather than something in our code.
3) There are four other groups with >135 species, although these are all at the end of the alphabet. I'm using 7 cores: my working hypothesis is that the last four groups maybe had more memory per core.
4) Hoverflies are the largest dataset among the groups I'm reading, in terms of numbers of records. This means the individual memory requirements of opening each model would be greater. However, leafSeedBeetles are in the middle of the range.

**Comment by AugustT** [comment]
@mlogie

**Comment by AugustT** [comment]
@drnickisaac It sounds like you think the fact that you are parallel reading is causing memory problems. This seems very likely as your 4gb memory is shared between your 7 cores + we know we have memory problems elsewhere. The simple solution/test would be to run one of these groups using fewer cores. Effectively that is what you are doing when you run serially, so this again re-enforces the need for users to be able to specify memory needs in datalabs. @mlogie does this seem reasonable to you? Additionally can you update us on your plan for the corrupted files?

**Comment by drnickisaac** [comment]
The corrupted files should be a separate issue.
I concur with @AugustT that it's something to do with the parallelisation. For example:
```
  filterDat_Gr <- (parallel::mclapply(sampDat,
                                                wrappeR::applyClipping,
                                                clipBy = "group",
                                                mc.cores = parallel::detectCores() - N))  
```
This fails for every value of `N`. Crucially, though, it's a different set of elements within `filterDat_Gr` that are corrupted. In this case I can do without the parallelization. For the reading in job we will certainly need more memory.

**Comment by robboyd** [comment]
Could still be memory ofc but for context we also got this error 
`#Error in load(paste0(indata, species, ".rdata")) : 
#  empty (zero-byte) input `file`
for for some hoverflies without parallelisation

**Comment by drnickisaac** [comment]
Ooh - how many. Let me guess: was it somewhere near 39?
And did you get the same for LeafSeedBeetles?
So maybe the two problems are related.
I've just noticed that some hoverfly and leafseedbeetles do have this in their data: `Error in load(paste0(indata, species, ".rdata")) : \n  empty (zero-byte) input file\n`

So obvious first steps:
1) @robboyd Rob could you write a catch for this error. If encountered then skip to the next species.
2) @mlogie to look at whether there are zero byte models on the object store.

This has sucked up a heap of my time already and I need to move on.

**Comment by drnickisaac** [comment]
The more I learn, the more I think @AugustT  is right. 
The example I presented above (`filterDat_Gr <- (parallel::mclapply(sampDat,`) no longer throws a wobbly if the data going in have no corruptions. There's just too much data to pick out the odd dodgy ones.
So the answer is clear: some good routines to catch the exceptions.

**Comment by mlogie** [comment]
Thanks @drnickisaac @AugustT thanks for bringing this to my attention.  I will loop through the files and try to work out what's happened.  If there was an issue with the upload I will overwrite the failed files with the original files.  Hopefully it was just confined to a small group, but I will try to loop over all groups to check if there were other errors

**Comment by drnickisaac** [comment]
Having ironed out other issues in wrappeR, I am confident now this is a memory limitation issue, not anything to do with the Object store. Specifically, when running the newest version of Caddisfly models, I get all species models out fine when `parallel=FALSE` but only a subset when `parallel=TRUE`.
So we could close this issue because it is a special case of #48 
@AugustT  I'll leave it open for now but part of the sprint: that way you can express a view and then get an easy win for the sprint.

**Comment by AugustT** [comment]
memory problem, closing
